lib/email_alert_title_builder.rb

```diff
@@ -0,0 +1,68 @@
+class EmailAlertTitleBuilder
+  def self.call(*args)
+    new(*args).call
+  end
+
+  def initialize(filter:, subscription_list_title_prefix:, facets:)
+    @filter = filter
+    @subscription_list_title_prefix = subscription_list_title_prefix
+    @facets = facets
+  end
+
+  def call
+    if facets.empty?
+      prefix
+    elsif facets.size == 1
+      (prefix + topic_names_sentence(facets.first)).upcase_first
+    elsif selected_facets.empty?
+      prefix
+    else
+      facet_sentences = selected_facets.map do |facet|
+        facet["facet_name"] + " of " + topic_names_sentence(facet)
+      end
+      (prefix + facet_sentences.to_sentence).upcase_first
+    end
+  end
+
+private
+
+  attr_reader :filter, :subscription_list_title_prefix, :facets
+
+  def topic_names(facet)
+    filter.fetch(facet["facet_id"], []).map { |key| choice_hash_by_key(facet, key)["topic_name"] }
+  end
+
+  def topic_names_sentence(facet)
+    topic_names(facet).to_sentence
+  end
+
+  def choice_hash_by_key(facet, key)
+    facet["facet_choices"].detect { |choice| choice["key"] == key }
+  end
+
+  def selected_facets
+    facets.select { |facet| filter[facet["facet_id"]].present? }
+  end
+
+  def plural_or_single
+    if filter.fetch(facets.first["facet_id"], []).length == 1
+      "singular"
+    else
+      "plural"
+    end
+  end
+
+  def prefix
+    if facets.empty?
+      subscription_list_title_prefix.to_s
+    elsif facets.size == 1
+      subscription_list_title_prefix[plural_or_single].to_s
+    elsif selected_facets.empty?
+      subscription_list_title_prefix.to_s
+    elsif subscription_list_title_prefix
+      "#{subscription_list_title_prefix}with "
```

Comment:
I think there's a missing space here 
/s/`"#{subscription_list_title_prefix}with "`/`"#{subscription_list_title_prefix} with "`

Reply:
In a sort of direct contrast to my comment above about creating a prefix and suffix method - there's lots of duplicate logic between the `call` and `prefix` - we are running mostly the same conditions so is there any value in splitting these methods up?

Also it looks like `facets.empty?` and `selected_facets.empty?` conditions return the same title regardless, so perhaps they could be merged into an `or`?

Reply:
It doesn't look ideal, but `subscription_list_title_prefix` includes a trailing space.

Reply:
Yeah, the logic used to be all in one method but we found it quite hard to read... I'm not sure if this is any better but we wanted to have a go at refactoring.

We could use an order for `facets.empty?` and `selected_facets.empty?` but we would have to include `facets.size < 1` for the `selected_facets` side because annoyingly the logic isn't quite as simple as an or. We can give it a go and see how it looks.